clang/lib/Parse/ParseDecl.cpp

```diff
@@ -4944,6 +4944,26 @@ void Parser::ParseStructDeclaration(
   }
 }
 
+// TODO: All callers of this function should be moved to
+// `Parser::ParseLexedAttributeList`.
+void Parser::ParseLexedCAttributeList(LateParsedAttrList &LAs, bool EnterScope,
+                                      ParsedAttributes *OutAttrs) {
+  assert(LAs.parseSoon() &&
+         "Attribute list should be marked for immediate parsing.");
+#ifndef NDEBUG
+  auto LangStd = getLangOpts().LangStd;
+  if (LangStd != LangStandard::lang_unspecified) {
+    auto Lang = LangStandard::getLangStandardForKind(LangStd).getLanguage();
+    assert(Lang == Language::C || Lang == Language::OpenCL);
```

Comment:
Was there a reason this couldn't be `!getLangOpts().CPlusPlus`?

Reply:
That doesn't sound equivalent. The language could also be objective-C or objective-C++.

Reply:
@hnrklssn Yes. The original assert didn't check for `Lang == Language::OpenCL` and that caused test failures due to the assertion failing.

Reply:
> That doesn't sound equivalent. The language could also be objective-C or objective-C++.
I think `getLangOpts().CPlusPlus` should set true when it is `Objective-C++`.

Reply:
> Isn't it? I think getLangOpts().CPlusPlus should set true when it is Objective-C++.
I'm not sure if that's the case but I'll check. But what about "Objective-C"? I didn't want this path taken for that so `!getLangOpts().CPlusPlus` would be true for Objective-C.

In general I think the approach of thinking that this function should be for C only is actually flawed and I'm going to look into removing this entirely after landing this.

Reply:
> In general I think the approach of thinking that this function should be for C only is actually flawed and I'm going to look into removing this entirely after landing this.

At that point, why not just call `ParseLexedAttributeList` and add a non-C++ case to `ParseLexedAttribute` instead of creating separate C versions that seem to serve basically the same purpose?

Reply:
So parsing a `.m` file without specifying a language seems to result in

```
(lldb) p LangOpts.LangStd
(clang::LangStandard::Kind) lang_gnu11
(lldb) p LangOpts.ObjC
(unsigned int) 1
(lldb) p LangOpts.C99
(unsigned int) 1
(lldb) p LangOpts.C11
(unsigned int) 1
(lldb) p LangOpts.C17
(unsigned int) 0
(lldb) p LangOpts.C23
(unsigned int) 0
(lldb) p LangOpts.CPlusPlus
(unsigned int) 0
(lldb) p clang::LangStandard::getLangStandardForKind(LangOpts.LangStd).getLanguage()
(clang::Language) C
```

So my assert doesn't do the right thing for Objective-C. For Objective-C my assert thinks its C 🤦‍♂️. So it seems like my assert would want to be `assert(!getLangOpts().CPlusPlus && !getLangOpts().ObjC)`. However, I think I'll remove this assert for now because I intend (provided I can make it to work with our internal code) to remove this code entirely in favor of re-using the late parsing code that's used for C++.

Reply:
>  However, I think I'll remove this assert for now because I intend (provided I can make it to work with our internal code) to remove this code entirely in favor of re-using the late parsing code that's used for C++

Ah so that's the same thing I suggested then. I thought you meant to just remove the assert. Overall I think this is a good idea 👍